File: boj/11052.cpp
#include <cstdio>
#include <cstring>
#define INF -987654321
#define max(a,b) a>b?a:b
int N,card[1001];
int cache[1001][1001];
int cost = INF;

int solve(int card,int num)
{
	if(card > N) return 0;
	if(card == N) return card[num];
	int &ret = cache[card][num];
	if(ret != -1)
		return ret;
	for(int i=0; i<N; i++)
		ret = max()
}

int main()
{
	memset(cache,-1,sizeof(cache));
	scanf("%d",&N);
	for(int i=0; i<N; ++i)
	{
		scanf("%d",&card[i]);
	}
	printf("%d",solve(0,0));
}

File: boj/python/11052.py
n = int(input())
cards = list(map(int, input().split()))

cache = [0] * 1001
cache[1] = cards[0]

for i in range(2, n+1):
	for j in range(1, i+1):
		cache[i] = max(cache[i], cache[i-j] + cards[j-1])

print(cache[n])